File: .idea/app/src/main/java/com/chevstrap/rbx/AppDirectories/CommonAppData.java
package com.chevstrap.rbx.AppDirectories;

import android.content.pm.PackageManager;
import com.chevstrap.rbx.App;

import java.io.File;

public abstract class CommonAppData {
    public abstract String getExecutablePackage();

    public CommonAppData() {}

    public String getExecutablePath() {
        try {
            String robloxDir = getRobloxDirectory();
            if (robloxDir != null) {
                return new File(robloxDir, "appData").getAbsolutePath();
            }
        } catch (Exception e) {
            App.getLogger().writeException("CommonAppData:getExecutablePath", e);
        }
        return null;
    }

    public String getRobloxDirectory() {
        String packageName = getExecutablePackage();
        if (packageName == null) return null;

        try {
            String dataDir = App.getAppContext()
                    .getPackageManager()
                    .getApplicationInfo(packageName, 0)
                    .dataDir;

            if (dataDir != null) {
                return new File(dataDir, "files").getAbsolutePath();
            } else {
                App.getLogger().writeLine("CommonAppData:getRobloxDirectory", "dataDir is null");
            }
        } catch (PackageManager.NameNotFoundException e) {
            App.getLogger().writeLine("CommonAppData:getRobloxDirectory", "failed getting package");
            App.getLogger().writeException("CommonAppData:getRobloxDirectory", e);
        }
        String fallback = App.getAppContext().getFilesDir().getAbsolutePath()
                .replace("/" + App.getAppContext().getPackageName() + "/", "/" + packageName + "/");
        return fallback + "/";
    }
}

File: app/src/main/java/com/chevstrap/rbx/appDirectories/CommonAppData.kt
package com.chevstrap.rbx.appDirectories

import android.content.pm.PackageManager
import com.chevstrap.rbx.App
import java.io.File

abstract class CommonAppData {
    abstract val executablePackage: String?
    val executablePath: String?
        get() {
            try {
                val robloxDir = this.robloxDirectory
                if (robloxDir != null) {
                    return File(robloxDir, "appData").absolutePath
                }
            } catch (e: Exception) {
                App.logger.writeException("CommonAppData::getExecutablePath", e)
            }
            return null
        }

    val robloxDirectory: String?
        get() = getRobloxDirectory(includeFiles = true)

    fun getRobloxDirectory(includeFiles: Boolean): String? {
        val packageName = this.executablePackage ?: return null

        try {
            val dataDir = App.appContext!!
                .packageManager
                .getApplicationInfo(packageName, 0)
                .dataDir

            if (dataDir != null) {
                return if (includeFiles) {
                    File(dataDir, "files").absolutePath
                } else {
                    File(dataDir).absolutePath
                }
            } else {
                App.logger.writeLine(
                    "CommonAppData::getRobloxDirectory",
                    "dataDir is null"
                )
            }
        } catch (e: PackageManager.NameNotFoundException) {
            App.logger.writeLine(
                "CommonAppData::getRobloxDirectory",
                "failed getting package"
            )
            App.logger.writeException(
                "CommonAppData::getRobloxDirectory",
                e
            )
        }

        val fallback = App.appContext!!.filesDir.absolutePath
            .replace(
                "/" + App.appContext!!.packageName + "/",
                "/$packageName/"
            )

        return if (includeFiles) {
            "$fallback/"
        } else {
            fallback.removeSuffix("/files")
        }
    }
}
